File: python/word-count/word_count.py
import re


quote_pattern = re.compile("^'.*'$")
def strip_quotes(word):
    if quote_pattern.match(word):
        return word[1:-1]
    else:
        return word


def word_count(phrase):
    pattern = re.compile("[^a-zA-Z0-9']")
    phrase = pattern.sub(' ', phrase)
    phrase = phrase.lower()
    words = phrase.split()
    words = [strip_quotes(w) for w in words]
    d = {}
    for word in words:
        if word not in d:
            d[word] = 1
        else:
            d[word] += 1
    return d

File: ruby/word-count/word_count.rb
# Count the number of occurences of each word in a phrase
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    punctuation_pattern = %r{[`~!@#\$%^&*\(\)\-_=+\[\]\\\{\}\|;:,.<>/?]}
    clean_phrase = @phrase.gsub(punctuation_pattern, ' ')
    words = clean_phrase.split(' ').map(&:downcase)
                        .map { |w| Surround.strip_surrounding_quotes(w) }
    words.each_with_object(Hash.new(0)) do |word, counts|
      counts[word] += 1
    end
  end
end

# Helper class for stripping surrounding strings
class Surround
  def self.strip_surrounding_quotes(word)
    word = strip_surrounding_string(word, '"') if surrounded_by(word, '"')
    word = strip_surrounding_string(word, "'") if surrounded_by(word, "'")
    # In case surrounded by single quotes first
    word = strip_surrounding_string(word, '"') if surrounded_by(word, '"')
    # We could recursively call for arbitrarily nested quotes, but let's not
    word
  end

  def self.strip_surrounding_string(string, surrounding)
    string.gsub(/^[#{surrounding}*]/, '').gsub(/#{surrounding}*$/, '')
  end

  def self.surrounded_by(string, surrounding)
    string.start_with?(surrounding) && string.end_with?(surrounding)
  end
end
